cozy/cozy-ui
# ✨ feat: transpilation

This is a work in progress. The goal is to transpile cozy-ui jsx and stylus files so that user of the library do not have to add something to their webpack config, thus improving usability of the library. This should also improve performance as the host webpack would have less to do.

## What's been done

JS files
* JSX files are transpiled into the `transpiled/` directory via `yarn transpile`

CSS
* A CSS stylesheet is extracted from the stylus files imported
by the JSX components during the `yarn transpile` phase. It should be imported by the host application.
* The CSS module classes are inserted into the components by the `css-module-transform plugin`

Polyfills
* Handled by babel through the [transform-runtime plugin](https://babeljs.io/docs/plugins/transform-runtime/)

Icons
* Bundled into a sprite just before transpilation. The application will have to add the sprite manually

## What's next

* Test on a real app, being done in Banks

## Breaking changes

1. Stylesheet must be inserted manually
2. `IconSprite` must be inserted manually
3. Icons lose their `icon-` prefix in their filename 

Check this for more information : https://github.com/cozy/cozy-ui/pull/468/files#diff-e7ea890a8dc2db7dbffd36fdc3d2226b

## How to test

On cozy-ui side

```
yarn link
yarn transpile --watch # will create the icon-sprite and babel watch
```

On app side

```
yarn link cozy-ui
```

You will need to add the icon sprite manually in the app (it was previously inserted by the svg-loader). [Check the example](https://github.com/cozy/cozy-ui/pull/468/files#diff-9777d1d36879cbac443b9da130b95b67R184).


## On polyfills

From the [documentation](https://babeljs.io/docs/plugins/transform-runtime/)

> This plugin is recommended in a library/tool.

> Babel uses very small helpers for common functions such as _extend. By default this will be added to every file that requires it. This duplication is sometimes unnecessary, especially when your application is spread out over multiple files.

> This is where the transform-runtime plugin comes in: all of the helpers will reference the module babel-runtime to avoid duplication across your compiled output. The runtime will be compiled into your build.
